### PR TITLE
Deferred acks for papertrail, socket, statsd sinks

### PR DESCRIPTION
**Pull request by spencergilbert**
- enhancement(papertrail sink,socket sink, statsd sink): Defer finalization until after processing in TcpSink
- enhancement(socket sink, statsd sink): Defer finalization until after processing in UdpSink
- enhancement(socket sink, statsd sink): Defer finalization until after processing in UnixSink

Closes: #13264
Closes: #13268
Closes: #13269

### CODE REVIEW

**Review by neuronull — APPROVED** [review]

File: src/sinks/util/udp.rs [hunk 1]
```diff
@@ -287,10 +287,9 @@ where
...
                 let byte_size = event.size_of();
 
-                self.acker.ack(1);
-
                 self.transformer.transform(&mut event);
 
+                let finalizers = event.take_finalizers();
                 let mut bytes = BytesMut::new();
                 if encoder.encode(event, &mut bytes).is_err() {
                     continue;
```
Comment (bruceg) [inline src/sinks/util/udp.rs:295]: I think you'll need to update the acker here…

File: src/sinks/util/udp.rs [hunk 2]
```diff
@@ -314,6 +313,9 @@ where
                         break;
```
Comment (bruceg) [inline src/sinks/util/udp.rs:313]: …and here.

File: src/sinks/util/tcp.rs [hunk 3]
```diff
@@ -291,6 +291,9 @@ where
             let _open_token = OpenGauge::new().open(|count| emit!(ConnectionOpen { count }));
 
             let mut mapped_input = stream::once(ready(item)).chain(&mut input).map(|event| {
+                drop(event.finalizers);
+                self.acker.ack(1);
+
```
Comment (bruceg) [inline src/sinks/util/tcp.rs:296]: I think this pattern (not your addition, but the existing setup) is questionable. If I'm reading right, this effectively does `sink.send_all(once(ready(item)))`, which means to send a stream containing a single item, followed by the stuff in the `chain` closure. I think that should be the same as just `sink.send(item)` followed by the rest of it, which makes for much simpler logic. Does that make sense?

File: src/sinks/util/unix.rs [hunk 4]
```diff
@@ -176,7 +178,15 @@ where
             let _open_token = OpenGauge::new().open(|count| emit!(ConnectionOpen { count }));
 
             let result = match sink
-                .send_all_peekable(&mut (&mut input).map(|item| item.item).peekable())
+                .send_all_peekable(
+                    &mut (&mut input)
+                        .map(|item| {
+                            drop(item.finalizers);
+                            self.acker.ack(1);
+                            item.item
```
Comment (bruceg) [inline src/sinks/util/unix.rs:186]: I think this is acking the item before it's sent and not after. Having said that, I'm not sure how to easily make this ack after.

**Review by bruceg — CHANGES_REQUESTED** [review]

File: src/sinks/util/udp.rs (hunk 1 above)
Comment (spencergilbert) [reply src/sinks/util/udp.rs:295]: So this didn't happen in your `console` commit that did just the deferral, but shouldn't we also drop the finalizers as we're acking here - later on the status updates happen in these stops, but if we hit these paths we'll also end up skipping the single `drop` that happens further down in the loop.

File: src/sinks/util/tcp.rs (hunk 3 above)
Comment (spencergilbert) [reply src/sinks/util/tcp.rs:296]: Maybe? Are you referring to just ln293 or the literal `sink.send_all` that happens down at ln311?

Comment (bruceg) [reply src/sinks/util/tcp.rs:296]: The whole thing, lines 293 through 314.

Comment (spencergilbert) [reply src/sinks/util/tcp.rs:296]: 028ce2a3a92335e9c36fe29f3aa3344614b70aee

Comment (spencergilbert) [reply src/sinks/util/tcp.rs:296]: Coming back to this after some other work - I'm not sure I grok the change. Do you have time today to pair on it really quickly?

Comment (bruceg) [reply src/sinks/util/tcp.rs:296]: Sure.

Comment (spencergilbert) [reply src/sinks/util/tcp.rs:296]: Recapping here - we misunderstood what was happening in the code here, and it should be fine to move forward

File: src/sinks/util/unix.rs (hunk 4 above)
Comment (bruceg) [reply src/sinks/util/unix.rs:186]: @jszwedko and I discussed this and agreed to push the remaining work for it to a separate issue to allow this one to close out. I'll write up that issue shortly.

**Review by bruceg — APPROVED** [review]

**Pull request merged**